backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,233 @@
+"""
+Creates a new DatasetVersion to update metadata across dataset artifacts
+"""
+import json
+import logging
+import os
+from typing import Dict
+
+import scanpy
+import tiledb
+from rpy2.robjects import StrVector
+from rpy2.robjects.packages import importr
+
+from backend.common.utils.corpora_constants import CorporaConstants
+from backend.common.utils.tiledb import consolidation_buffer_size
+from backend.layers.business.business import BusinessLogic
+from backend.layers.common.entities import (
+    CollectionVersionId,
+    DatasetArtifactType,
+    DatasetConversionStatus,
+    DatasetProcessingStatus,
+    DatasetStatusKey,
+    DatasetValidationStatus,
+    DatasetVersion,
+    DatasetVersionId,
+)
+from backend.layers.persistence.persistence import DatabaseProvider
+from backend.layers.processing.logger import configure_logging
+from backend.layers.processing.process_download import ProcessDownload
+from backend.layers.thirdparty.s3_provider import S3Provider
+from backend.layers.thirdparty.uri_provider import UriProvider
+
+configure_logging(level=logging.INFO)
+
+base = importr("base")
+seurat = importr("SeuratObject")
+
+
+class DatasetMetadataUpdate(ProcessDownload):
+    def __init__(
+        self, business_logic: BusinessLogic, artifact_bucket: str, cellxgene_bucket: str, datasets_bucket: str
+    ) -> None:
+        super().__init__(business_logic, business_logic.uri_provider, business_logic.s3_provider)
+        self.artifact_bucket = artifact_bucket
+        self.cellxgene_bucket = cellxgene_bucket
+        self.datasets_bucket = datasets_bucket
+
+    def update_h5ad(
+        self,
+        h5ad_uri: str,
+        original_dataset_version: DatasetVersion,
+        key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update_dict: Dict[str, str],
+    ):
+        h5ad_filename = self.download_from_source_uri(
```

Comment:
Downloading and uploading the updated dataset uses very little CPU and memory. We should run these steps in a separate thread to make the most out of the machines.

Reply:
I'd also recommend using a Processing pool to avoid running over resources.